Source/Dafny/AST/DafnyAst.cs

```diff
@@ -6142,6 +6211,51 @@ public BoundVar(IToken tok, string name, Type type)
       Contract.Requires(name != null);
       Contract.Requires(type != null);
     }
+
+    public bool CurrentTypeAssumedToBeCompilable { get; set; } = false;
+
+    public void AssumeCompilableType(Type secondaryType) {
+      Contract.Assert(!CurrentTypeAssumedToBeCompilable);
+      otherType = secondaryType;
+      AssumeCompilableTypeIfAny();
+    }
+
+    public void AssumeCompilableTypeIfAny() {
+      if (!CurrentTypeAssumedToBeCompilable && otherType != null) {
+        (otherType, type) = (type, otherType);
+        CurrentTypeAssumedToBeCompilable = true;
+      }
+    }
+
+    public void AssumeOriginalType() {
+      if (CurrentTypeAssumedToBeCompilable) {
+        Contract.Assert(otherType != null);
+        (type, otherType) = (otherType, type);
+        CurrentTypeAssumedToBeCompilable = false;
+      }
+    }
+
+    public void AcceptOriginalTypeAssumption() {
+      if (CurrentTypeAssumedToBeCompilable) {
+        Contract.Assert(otherType != null);
+        (type, otherType) = (otherType, null);
+        CurrentTypeAssumedToBeCompilable = false;
+      } else {
+        otherType = null;
+      }
+    }
+
+    public Type CompilableType => CurrentTypeAssumedToBeCompilable ? Type : otherType;
+    public Type OriginalType => CurrentTypeAssumedToBeCompilable ? otherType : Type;
+
+    public bool NeedsToVerifyOriginalTypeInference() {
+      return otherType != null && !OriginalType.IsCompilable();
+    }
+    public bool CompilableTypeAssumptionIsNotTrivial() {
+      return otherType != null && !otherType.Equals(Type, true);
+    }
+
+    private Type otherType = null;
```

Comment:
Indeed, these types are swapped sometimes, but here is something to note:
`otherType` is not the publicly accessible field. The fields that should be publicly accessed to access stable types are are `CompilableType``and `OriginalType`.

The rationale because of the swapping is that I did not want to change the "scope" mechanism that is being used by the resolver: `scope.Push(v.Name, v);`. It pushes both the variable's name and the variable itself (with is `.Type`) in the scope.
However, I needed to _resolve_ the range in two settings: when the type of that variable is the type the user has specified (in case it's explicit), and when the type of that variable is the compilable supertype of the original type.
For the resolver, swapping enables resolution to not care about that is the context which defines the variable's type, just access `Type`.
I tried alternatively to create another bound variable with the other type, but it would then allocate a new compiled name for it, which was undesirable.

Reply:
I see. So there are three public fields: `Type`, `CompilableType` and `OriginalType`, and two backing field `type` and `otherType`, that work with the boolean `CurrentTypeAssumedToBeCompilable`.

How do you feel about letting `CompilableType` and `OriginalType` use a backing field, and implementing `Type` using `Type => CurrentTypeAssumedToBeCompilable ? CompilableType : OriginalType` ?

`OriginalType` could use `type` as its backing field.

--- 
I'm still trying to grasp why BoundVariable needs four state changing operations:
- `AssumeCompilableType`
- `AssumeCompilableTypeIfAny`
- `AssumeOriginalType`
- `AcceptOriginalTypeAssumption`

I would have hoped you only go from one state to another, but it seems like you go back and forth. 

```
        if (s.IsBindingGuard) {
          var exists = (ExistsExpr)s.Guard;
          AssumeOriginalTypeForBoundedVariables(exists!);
          foreach (var v in exists!.BoundVars) {
            ScopePushAndReport(scope, v, "bound-variable");
          }
        }
        dominatingStatementLabels.PushMarker();
        ResolveBlockStatement(s.Thn, codeContext);
        dominatingStatementLabels.PopMarker();
        scope.PopMarker();
        if (s.IsBindingGuard) {
          var exists = (ExistsExpr)s.Guard;
          AssumeCompilableTypeForBoundedVariables(exists!);
        }
```
What's the reason for the above code?

---

Would it be possible to use only two public types, `Type` and `SpecifiedType`, where `SpecifiedType` has its own backing field and is always the user specified type, and implement `BoundVariable.AssumeOriginalType` by saying `Type = SpecifiedType`
and `BoundVariable.AssumeCompilableType` with `Type = compilableType` ?

Reply:
> How do you feel about letting `CompilableType` and `OriginalType` use a backing field, and implementing `Type` using `Type => CurrentTypeAssumedToBeCompilable ? CompilableType : OriginalType` ?
> 
> `OriginalType` could use `type` as its backing field.

Thanks! That was helpful. I implemented it.

Reply:
> What's the reason for the above code?

Ok I managed to simplify it at the end, but here is the story because I think there is still more work to do in the future. I first came across the following case:

```dafny
if x :| P(x) {
  S(x);
}
```

Which actually uses an exists expression with a bound variable `x`, but further makes the "then branch" bound with that same bound variable. Hence, it's as if the "then body" was part of the "term expression" of the exists. However, after resolving the Exists expression, the resolver currently sets back the "compiled" type for the bound variable, if it exists, for the verifier later to verify that the ghost constraint holds.
So, I need to temporarily revert back to the "specified type" so that type inference in the then body uses the specified type, not the compilable type, and then I put the "compiled type" for later.

I just tried to ensure that the verifier uses the compilable type anyway, so that we don't need to switch back and forth (and leave the specified type for bound variables).

This lead me to discover the following inception challenge that would require more thinking now:

```dafny
ghost const zero := 0
type GhostNat = x : int | x >= zero
const S: set<int> := {-1, 0, 1};
type GhostNatInS = x : int | x in (set j: GhostNat | j in S && j >= 0)

function method FailIfNotGhostNatInS(i: GhostNatInS): int {
  if i < 0 then 1/0 else i
}

method Main() {
  var s := forall x: GhostNatInS | x in S :: FailIfNotGhostNatInS(x) == x;
  print s;
}
```
The challenging question is:

*Is the constraint of the type `GhostNatInS` ghost or not?*

If it is statically ghost, then the Dafny verifier will detect that the `forall` range cannot prove that `x is GhostNatInS`. If it is not statically ghost, then the Dafny verifier will verify that `j in S && j >= 0` implies that `j is GhostNat`, which is true, and then succeed, which is what we want.
However, "ghostness" is currently inferred statically from the subset constraint itself, and there is no check to determine if `set i: GhostNat | ...` is ghost, so it will infer it's compilable (because it has bounds). The constraint is currently [parsed as ghost](https://github.com/dafny-lang/dafny/pull/1702/files#diff-7d845d452486b0b59700ca5c133f2648f76d44616148081d51efffe32dbc6e7aL2674) anyway, so it's consistent with the fact that it should not find any kind of ghostness to infer that the constraint is compilable.

Note that if any bound variable with a ghost constraint was triggering its enclosing expression to be ghost, that means that the above example would not work because `s` would be ghost and the print statement would fail, and this PR expressedly aims at avoiding this abusive ghost inference.

To sum up, we have the following problem:
- The constraint of `GhostNatInS` is currently resolved as in a "ghost context", so that it tolerates if the bound variable has a ghost subset type.
- When checking whether the constraint is ghost or not, we don't infer that bound variables with a ghost subset type trigger the whole expression to be ghost. If we did the same algorithm would infer that every comprehension with a ghost subset type cannot be compiled, which this PR aims at avoiding.
- Thus, the algorithm thinks it can inline this constraint. In the example, above, and if we removed `j >= 0`  (test `git-issue-1604h.dfy`), it would throw an exception at run-time after the Dafny verifier would tell everything is ok

The current master avoid all these problems because it does not allow ghost constraints in compiled contexts, and because the constraint is immediately resolved to be compilable or not after resolution and (set j | ...)` does not have inferred bounds yet (bounds inference comes after resolution), it's marked as ghost. So, it's a kind of a hack that it works.

In this PR, I changed the way we parse the subset constraint, and assumed it's compilable, so that the verifier will verify such subset types. This mean it introduces one breaking change: It's no longer possible to have ghost comprehensions in the constraint of a subset type, they would need to be wrapped in a ghost predicate instead. I'll document that.

@RustanLeino, if you are able to follow, I would love your advice on this. It seems that it would be great to add the keywords "ghost" and/or "compiled" to the type to be able to enforce one or the other.